File: group1/112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        if not root.left and not root.right:
            return sum == root.val
        return self.hasPathSum(root.left, sum - root.val) or self.hasPathSum(root.right, sum - root.val)


class Solution:
    # bfs
    def hasPathSum(self, root: TreeNode, sum: int) -> bool:
        if not root:
            return False
        queue = [(root, root.val)]

        while queue:
            node, path = queue.pop(0)
            if not node.left and not node.right and path == sum:
                return True
            left, right = node.left, node.right
            if left:
                queue.append((left, path+left.val))
            if right:
                queue.append((right, path+right.val))

        return False

File: leetcode/group2/112.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {number[][]}
 */
var pathSum = function (root, targetSum) {
  const ret = [];
  const ans = [];

  const dfs = (node, targetSum) => {
    if (!node) {
      return;
    }

    const { left, right, val } = node;

    if (val === targetSum && !left && !right) {
      ans.push(val);
      ret.push(ans.slice());
      ans.pop();
      return;
    }

    if (left) {
      ans.push(val);
      dfs(left, targetSum - val);
      ans.pop();
    }
    if (right) {
      ans.push(val);
      dfs(right, targetSum - val);
      ans.pop();
    }
  };

  dfs(root, targetSum);

  return ret;
};

// improve
var pathSum = function (root, targetSum) {
  const ret = [];
  const ans = [];

  const dfs = (node, targetSum) => {
    if (!node) {
      return;
    }
    const { left, right, val } = node;
    ans.push(val);
    if (val === targetSum && !left && !right) {
      ret.push(ans.slice());
    }
    dfs(left, targetSum - val);
    dfs(right, targetSum - val);
    ans.pop();
  };

  dfs(root, targetSum);
  return ret;
};
